### ISSUE TITLE
Cyclic dependency between `webpack` and `webpack-cli` is not represented

### ISSUE THREAD

**Issue by Silic0nS0ldier** [aspect-build/rules_js#497]
Repro with https://github.com/Silic0nS0ldier/repro-rules-js-dep-graph-bug

When running outside of Bazel with PNPM (e.g. `cd web ; pnpm i ; pnpm exec webpack build`) dependencies are correctly resolved.
When running the equivalent in Bazel (e.g. `bazel build //web/src/webpack-testing:baz-min`) dependencies (specifically `webpack` from `webpack-cli`) cannot be resolved.

When inspecting `node_modules/.aspect_rules_js/webpack-cli@4.10.0...` no `webpack` symlink exists. It does in the PNPM equivalent.

**Comment by cgrindel** [comment]
@thesayyn may have a possible workaround for this issue. He will post here shortly.

**Comment by thesayyn** [comment]
TLDR; the workaround here is to use `--experimental_allow_unresolved_symlinks` to actually symlink.

this is due to the limitation of bazel as of today. when two packages depend on each other, bazel won't be able to build it. so as a workaround, we break them down into two with their closure dep graph. basically, you get an identical copy of `webpack` for `webpack-cli` and `webpack-cli` for `webpack`. they are not symlinked but hard copied.

**Comment by cgrindel** [comment]
@Silic0nS0ldier Does @thesayyn workaround address your issue? If so, can we close this issue?

**Comment by Silic0nS0ldier** [comment]
Hi all, sorry for the delay.

I've tried using `--experimental_allow_unresolved_symlinks`, I just end up with different errors.

Repro repo:
```
bazel build //web/src/webpack-testing:baz-min --experimental_allow_unresolved_symlinks
Starting local Bazel server and connecting to it...
INFO: Analyzed target //web/src/webpack-testing:baz-min (158 packages loaded, 644 targets configured).
INFO: Found 1 target...
ERROR: /Users/mele/source/private/rules_js-webpack-deps-bug_2022-10-06/web/src/webpack-testing/BUILD.bazel:3:20: Webpack web/src/webpack-testing/baz-min.js failed: (Exit 1): baz-min__js_binary.sh failed: error executing command bazel-out/darwin_arm64-opt-exec-2B5CBBC6/bin/web/src/webpack-testing/baz-min__js_binary.sh build

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module 'webpack-cli/package.json'
Require stack:
- /private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/120/execroot/__main__/bazel-out/darwin_arm64-opt-exec-2B5CBBC6/bin/web/src/webpack-testing/baz-min__js_binary.sh.runfiles/__main__/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/webpack/bin/webpack.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at runCli (/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/120/execroot/__main__/bazel-out/darwin_arm64-opt-exec-2B5CBBC6/bin/web/src/webpack-testing/baz-min__js_binary.sh.runfiles/__main__/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/webpack/bin/webpack.js:65:26)
    at Object.<anonymous> (/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/120/execroot/__main__/bazel-out/darwin_arm64-opt-exec-2B5CBBC6/bin/web/src/webpack-testing/baz-min__js_binary.sh.runfiles/__main__/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/webpack/bin/webpack.js:162:2)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/120/execroot/__main__/bazel-out/darwin_arm64-opt-exec-2B5CBBC6/bin/web/src/webpack-testing/baz-min__js_binary.sh.runfiles/__main__/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/webpack/bin/webpack.js'
  ]
}
Target //web/src/webpack-testing:baz-min failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 8.856s, Critical Path: 3.11s
INFO: 282 processes: 163 internal, 119 local.
FAILED: Build did NOT complete successfully
```

I also noticed `--experimental_allow_unresolved_symlinks` used in the scenario that the repro was pulled from ran into new issues where previously an installed would have succeeded. (failure in install hooks citing a `Cannot find module` error).

**Comment by Silic0nS0ldier** [comment]
Did some more investigation. If I run `bazel build //web:node_modules` and inspect the output, everything looks correct.

I also tested again with the latest release (v1.6.4). Still failing with same reported error (including with `--experimental_allow_unresolved_symlinks`).

**Comment by Silic0nS0ldier** [comment]
What I hadn't noticed before was when running with `--experimental_allow_unresolved_symlinks` is that by the time webpack is being run, the generated `node_modules` structure is littered with broken symlinks.

```
File attempting to import
/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/716/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/webpack/bin/webpack.js

Nearest node_modules
/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/716/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/webpack@5.74.0_bhdynqia3mcfj5mxwwuhn5skbe_webpack-cli@4.10.0/node_modules/

Contents of nearest node_modules
@types                          chrome-trace-event              glob-to-regexp                  neo-async                       webpack
@webassemblyjs                  enhanced-resolve                graceful-fs                     schema-utils                    webpack-cli
acorn                           es-module-lexer                 json-parse-even-better-errors   tapable                         webpack-sources
acorn-import-assertions         eslint-scope                    loader-runner                   terser-webpack-plugin
browserslist                    events                          mime-types                      watchpack

Symlink destination for "webpack-cli"
/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/webpack-cli@4.10.0_ke6kce33fn7czsdxd7je7n4eo4_webpack@5.74.0/node_modules/webpack-cli

What would have worked
/private/var/tmp/_bazel_mele/9c39448532a075544e3f467e6f03cca2/sandbox/darwin-sandbox/716/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/node_modules/.aspect_rules_js/webpack-cli@4.10.0_ke6kce33fn7czsdxd7je7n4eo4_webpack@5.74.0/node_modules/webpack-cli

Contents of nearest node_modules, for "webpack-cli"
@discoveryjs            colorette               cross-spawn             import-local            rechoir                 webpack-merge
@webpack-cli            commander               fastest-levenshtein     interpret               webpack-cli
```

Note that there is no reference back to `webpack` for the `webpack-cli` location. It seems like it got further without `--experimental_allow_unresolved_symlinks` if the `requireStack` if anything to go by.

**Comment by thesayyn** [comment]
> What I hadn't noticed before was when running with `--experimental_allow_unresolved_symlinks` is that by the time webpack is being run, the generated `node_modules` structure is littered with broken symlinks.

this sounds familiar. What version of bazel are you on? Could you try with 6.0.0rc1? 

@gregmagolan might have other ideas why this is not working.

**Comment by Silic0nS0ldier** [comment]
I can't see any `6.0.0-rc.1` release at present so I'll test with the latest v6 release at time of writing, [`6.0.0-pre.20221220.1`](https://github.com/bazelbuild/bazel/releases/tag/6.0.0-pre.20221020.1).

| OS | Arch | Default Configuration | `--experimental_allow_unresolved_symlinks` |
|---|---|---|---|
| macOS 12.6.1 | `aarch64` | Crash, no "webpack" | Crash, no "webpack" |
| macOS 12.6.0 | `amd64` | Crash, no "webpack" | Crash, no "webpack" |
| Ubuntu 22.04.1 LTS | `amd64` | Crash, no "webpack" | Crash, no "webpack" |
| Windows 11 (out of curiosity) | `amd64` | Crash, unrelated | Crash, unrelated |

**Comment by gregmagolan** [comment]
The `6.0.0rc` release isn't on the Bazel releases page but if you set it in your `.bazelversion` and are using bazelisk, bazelisk knows how to download it, https://github.com/aspect-build/rules_js/blob/4a039ea6701eac6471f2c147041cd448181c4976/.bazelversion#L1.

However, there shouldn't be that much of a difference between 6.0.0r1 and 6.0.0-pre.20221220.1

**Comment by gregmagolan** [comment]
I'll TAL at your repro

**Comment by gregmagolan** [comment]
Played around a bit and I think I see how a bug manifested. Bazel doesn't support cyclic dependencies between targets so there was some magic done with target "references" to support it between npm deps. It may be that you're the first to stumble on this case.

We do have an example in rules_js of running webpack-cli instead of web successfully, https://github.com/aspect-build/rules_js/blob/4a039ea6701eac6471f2c147041cd448181c4976/examples/webpack_cli/BUILD.bazel#L29. The difference there is that we're using the webpack-cli bin and in that case the webpack-cli dep on webpack does show up in the virtual store. In your case you're using the webpack bin and the that dep doesn't show up which looks like a bug. I'll dig around a bit more in the code to find the root cause.

**Comment by gregmagolan** [comment]
We also have a ruleset for webpack https://github.com/aspect-build/rules_webpack that comes with a useful webpack_devserver rule. I'll track down the bug in an case.

**Comment by gregmagolan** [comment]
Fix will look something like this https://github.com/aspect-build/rules_js/pull/582; tho in that shape it causes another failure on rules_js so have to sort that out. But it does resolve the problem in your repro. You can try it out with the commit from the PR:

```$ git diff
diff --git a/WORKSPACE.bazel b/WORKSPACE.bazel
index e22dbaa..21d679e 100644
--- a/WORKSPACE.bazel
+++ b/WORKSPACE.bazel
@@ -2,9 +2,9 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
     name = "aspect_rules_js",
-    sha256 = "02da7c546cd70751b7375b7e95f7298709f5a33fda2863c2b68dac7745bca66d",
-    strip_prefix = "rules_js-1.6.4",
-    url = "https://github.com/aspect-build/rules_js/archive/refs/tags/v1.6.4.tar.gz",
+    sha256 = "c114ee6370d9814aa91f616b6198318cfa43c73512710785b6f65bfca4a68c36",
+    strip_prefix = "rules_js-6e084e4b00ea2000d59bfc56ecd06a27bd44ea47",
+    url = "https://github.com/aspect-build/rules_js/archive/6e084e4b00ea2000d59bfc56ecd06a27bd44ea47.tar.gz",
 )
 
 load("@aspect_rules_js//js:repositories.bzl", "rules_js_dependencies")```

**Comment by gregmagolan** [comment]
Fix is all green in https://github.com/aspect-build/rules_js/pull/582. I'll cut a rules_js release after landing tomorrow.

**Comment by Silic0nS0ldier** [comment]
Thanks @gregmagolan, got to love those dyadic circular dependencies.

Results from latest test battery with v6.

```
| OS | Arch | Default Configuration | `--experimental_allow_unresolved_symlinks` |
|---|---|---|---|
| macOS 12.6.1 | `aarch64` | Success-ish | Success-ish |
| macOS 12.6.0 | `amd64` | Success-ish | Success-ish |
| Ubuntu 22.04.1 LTS | `amd64` | Success-ish | Success-ish |
```

Success-ish = The expected "invalid webpack configuration" crash.

Also ran some tests under v5, which as you'll soon see...

```
| OS | Arch | Default Configuration | `--experimental_allow_unresolved_symlinks` |
|---|---|---|---|
| macOS 12.6.1 | `aarch64` | Success-ish | Crash, no "webpack-cli/package.json" |
```

...apparently has a bug with the `--experimental_allow_unresolved_symlinks` flag.

**Comment by gregmagolan** [comment]
Yes. In Bazel 5.3 the unresolved symlinks feature wasn't stabilized yet and can't be used with rules_js, but should be able to use Bazel 5.3 without the flag set.

The fix is out in the v1.6.6 release https://github.com/aspect-build/rules_js/releases/tag/v1.6.6
